lib/ansible/galaxy/api.py

```diff
@@ -384,6 +395,9 @@ def _call_galaxy(self, url, args=None, headers=None, method=None, auth_required=
                             method=method, timeout=self._server_timeout, http_agent=user_agent(), follow_redirects='safe')
         except HTTPError as e:
             raise GalaxyError(e, error_context_msg)
+        except (URLError, OSError):
+            # Other exceptions we may want to retry, these may be filtered further by should_retry_error
+            raise
         except Exception as e:
             raise AnsibleError("Unknown error when attempting to call Galaxy at '%s': %s" % (url, to_native(e)))
```

Comment:
I'm guessing most of the test failures are due to letting an unhandled OSError/URLError trickle up.

I found a draft from looking at this issue a couple months ago, where I passed the original exception through so the callers wouldn't need to all handle that.
```diff
diff --git a/lib/ansible/galaxy/api.py b/lib/ansible/galaxy/api.py
index 8dea804953..eb07b2a11e 100644
--- a/lib/ansible/galaxy/api.py
+++ b/lib/ansible/galaxy/api.py
@@ -33,7 +33,7 @@ from ansible.utils.path import makedirs_safe
 
 display = Display()
 _CACHE_LOCK = threading.Lock()
-COLLECTION_PAGE_SIZE = 100
+COLLECTION_PAGE_SIZE = 1
 RETRY_HTTP_ERROR_CODES = [  # TODO: Allow user-configuration
     429,  # Too Many Requests
     520,  # Galaxy rate limit error code (Cloudflare unknown error)
@@ -48,7 +48,14 @@ def cache_lock(func):
     return wrapped
 
 
-def is_rate_limit_exception(exception):
+def is_rate_limit_exception(exception):  # FIXME
+    if (
+        isinstance(exception, AnsibleError)
+        and exception.orig_exc is not None
+        and isinstance(exception.orig_exc, OSError)
+    ):
+        return True
+
     # Note: cloud.redhat.com masks rate limit errors with 403 (Forbidden) error codes.
     # Since 403 could reflect the actual problem (such as an expired token), we should
     # not retry by default.
@@ -385,7 +392,10 @@ class GalaxyAPI:
         except HTTPError as e:
             raise GalaxyError(e, error_context_msg)
         except Exception as e:
-            raise AnsibleError("Unknown error when attempting to call Galaxy at '%s': %s" % (url, to_native(e)))
+            user_error = "Unknown error when attempting to call Galaxy at '%s': %s" % (url, to_native(e))
+            if isinstance(e, OSError) and e.args and e.args[0] == "The read operation timed out":
+                raise AnsibleError(user_error, orig_exc=e)
+            raise AnsibleError(user_error)
```

(the page size was just hardcoded to 1 to help reproduce the issue - I don't know how people are hitting this normally with Galaxy...)

Reply:
Nice, I'll look at adjusting my PR.  I had thought of doing something similar, just hadn't gotten there. Since `AnsibleError` is treated more nicely, than other arbitrary exceptions.